runtime/lib/ttnn/operations/eltwise/tertiary/where.cpp

```diff
@@ -0,0 +1,24 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "where.h"
+#include "tt/runtime/detail/ttnn.h"
+#include "tt/runtime/ttnn/operations/utils.h"
+
+namespace tt::runtime::ttnn::operations::where {
+
+void run(const ::tt::target::ttnn::WhereOp *op, ProgramContext &context) {
+  ProgramTensorPool &tensorPool = context.tensorPool;
+
+  ::tt::tt_metal::MemoryConfig outputMemoryConfig =
+      utils::createMemoryConfig(op->out());
+  const ::ttnn::Tensor &in = tensorPool.at(op->in()->global_id());
+
+  ::ttnn::Tensor out =
+      ::ttnn::where(in, op->pred(), op->on_true(), op->on_false(),
```

Comment:
What are we doing here? Aren't op->on_true(), op->on_false() tensor ref pointers? 
```
table WhereOp {
  pred: tt.target.TensorRef;
  on_true: tt.target.TensorRef;
  on_false: tt.target.TensorRef;
  output: tt.target.TensorRef;
}
```
Also I think the correct way to implement this is to create an eltwise ternary file and add this op here, since ttnn models this op as an eltwise ternary op.

Reply:
Also please add a silicon test for this under Silicon/TTNN

Reply:
> The reason this doesn't break anything is probably because this file wasn't added to runtime/lib/ttnn/operations/CMakeLists.txt

Oh, that explains [this](https://github.com/tenstorrent/tt-mlir/pull/852#discussion_r1830074361) :) 

> Also I think the correct way to implement this is to create an eltwise ternary file and add this op there, since ttnn models this op as an eltwise ternary op.

@jnie-TT Can you check [this](https://github.com/tenstorrent/tt-mlir/pull/852#discussion_r1819112960) comment please? In short, Metal currently also has a separate file for where op since it has different parameters from other ternary ops.

Reply:
@mrakitaTT Yeah we should create a new file, and we should encapsulate the op in the ternary namespace, possibly also modelling this op as ternary in the flatbuffer schema.

Reply:
> @mrakitaTT Yeah we should create a new file, and we should encapsulate the op in the ternary namespace, possibly also modelling this op as ternary in the flatbuffer schema and maybe in the compiler as well.

I think that unless we can find two or more tertiary ops that behave the same way like we had for binary and unary ops, there is no point of creating a common infrastructure for them and we should keep them separate like I did with where op and be treated like a standalone op.

Reply:
> Yeah we should create a new file, and we should encapsulate the op in the ternary namespace

Yeah, Usman moved it into separate file and made a ternary folder, but you are right he misnamed it into `tertiary` and didn't update the namespace, good catch on those, @uazizTT can you please fix?

> possibly also modelling this op as ternary in the flatbuffer schema and maybe in the compiler as well

@jnie-TT As you can see in that comment I didn't suggest Usman to make other changes yet because Metal currently keeps `where` op in a separate file and there is no connection with other ternary ops except it being in the same folder as them. It also has a different interface than other ternary ops, so I don't see much that we can do in the compiler yet. I think the best we can currently do is to match Metal structure and keep where op separate from other ternary ops, and when they make a change we can make a change too. But I'll let you and Kyle decide of course :)

Reply:
Where op should be a standalone file in runtime because it's a standalone file in ttnn. 
I think the goal is to model as closely to ttnn as possible, so we should model this closely to ttnn from the start instead of having to merge the ops later. Even if we only have 1 ternary op we should model it as a ternary op instead of a standalone op. We did the same for reduce ops, where initially we had only reduce sum, but we still modelled it in the bracket of reduce op.

Reply:
@mrakitaTT I agree that we should make it a separate file than other ternary ops. But I think that in at least in the ttnn program flatbuffer schema we should model it as a ternary op. For example unary.hpp and unary_composite.hpp are separate files, but we still model them as eltwise unary ops in flatbuffer. If we need extra parameters we can add an interface for that. Looking at the interface, it's not that much different from other ternary ops:
```
Tensor WhereOperation::invoke(
    uint8_t queue_id,
    const Tensor& predicate,
    const Tensor& value_true,
    const Tensor& value_false,
    const std::optional<MemoryConfig>& output_mem_config,
    std::optional<Tensor> output_tensor) {

    return ternary_utils::where_impl(queue_id, predicate, value_true, value_false, output_mem_config.value_or(predicate.memory_config()), output_tensor);
}
```
Other ternary ops follow roughly the same structure, while some may require an extra float argument.

Reply:
> I think that unless we can find two or more tertiary ops that behave the same way like we had for binary and unary ops, there is no point of creating a common infrastructure for them and we should keep them separate like I did with where op and be treated like a standalone op.

@uazizTT I agree about treating where op separately for now, but as for the common infrastructure part I think that as soon as we have the first ternary op we should make a support for it in a general way so that other ternary ops can be added in the same file. Same as what @jnie-TT said in the comment above, we are overlapping with comments right now :)

Reply:
If it's too much work or it's unclear how this op will interface with other ternary ops in some common infra, then yeah we can add it standalone for now, but please leave a note that this should be considered when creating an infra for ternary.

Reply:
> I agree that we should make it a separate file than other ternary ops. But I think that at least in the ttnn program flatbuffer schema we should model it as a ternary op. For example unary.hpp and unary_composite.hpp are separate files, but we still model them as eltwise unary ops in flatbuffer. If we need extra parameters we can add an interface for that.

@jnie-TT Sure, makes sense

Reply:
Thanks guys for the discussion! I'm trying to make runtime as close to TTNN as possible and limiting the amount of workarounds when needed. I know it can be cumbersome at times but as we bring up more ops it'd be desirable if we can keep it organized in this manner.

Reply:
@jnie-TT np, so just to make clear, is this the set of required changes that we converged to?
- Rename folder and include guard to use `ternary` instead of `tertiary`
- Fix the namespace
- Create common ternary ops interface in flat buffer and then extend interface for where op from it